pkg/retriever/bitswapretriever.go

```diff
@@ -1,29 +1,194 @@
+// Selectors are used to travers the dag to make sure the CARs for bitswap match graphsync
+// Note: this is a tradeoff over go-merkledag for traversal, cause selector execution is slow. But the solution
+// is to improve selector execution, not introduce unpredictable encoding.
+type BitswapRetriever struct {
+	bstore       MultiBlockstore
+	routing      IndexerRouting
+	blockService blockservice.BlockService
 }
 
+// NewBitswapRetrieverFromHost constructs a new bitswap retriever for the given libp2p host
+func NewBitswapRetrieverFromHost(ctx context.Context, host host.Host) *BitswapRetriever {
+	bstore := bitswaphelpers.NewMultiblockstore()
+	routing := bitswaphelpers.NewIndexerRouting()
+	bsnet := network.NewFromIpfsHost(host, routing)
+	bitswap := client.New(ctx, bsnet, bstore)
+	bsnet.Start(bitswap)
+	bsrv := blockservice.New(bstore, bitswap)
+	return NewBitswapRetrieverFromDeps(bsrv, routing, bstore)
+}
+
+// NewBitswapRetrieverFromDeps is primarily for testing, constructing behavior from direct dependencies
+func NewBitswapRetrieverFromDeps(bsrv blockservice.BlockService, routing IndexerRouting, bstore MultiBlockstore) *BitswapRetriever {
+	return &BitswapRetriever{
+		bstore:       bstore,
+		routing:      routing,
+		blockService: bsrv,
+	}
+}
+
+// Retrieve initializes a new bitswap session
 func (br *BitswapRetriever) Retrieve(ctx context.Context, request types.RetrievalRequest, events func(types.RetrievalEvent)) types.CandidateRetrieval {
-	return &bitswapRetrieval{br, ctx, request, events}
+	return &bitswapRetrieval{br, blockservice.NewSession(ctx, br.blockService), ctx, request, events}
 }
 
 type bitswapRetrieval struct {
 	*BitswapRetriever
-	ctx     context.Context
-	request types.RetrievalRequest
-	events  func(types.RetrievalEvent)
+	bsGetter blockservice.BlockGetter
+	ctx      context.Context
+	request  types.RetrievalRequest
+	events   func(types.RetrievalEvent)
 }
 
+// RetrieveFromCandidates retrieves via go-bitswap backed with the given candidates, under the auspices of a fetcher.Fetcher
 func (br *bitswapRetrieval) RetrieveFromCandidates(candidates []types.RetrievalCandidate) (*types.RetrievalStats, error) {
-	return nil, errors.New("not implemented")
+	phaseStartTime := time.Now()
+	// this is a hack cause we aren't able to track bitswap fetches per peer for now, so instead we just create a single peer for all events
+	bitswapCandidate := types.NewRetrievalCandidate(peer.ID(""), br.request.Cid, metadata.Bitswap{})
+	br.events(events.Started(br.request.RetrievalID, phaseStartTime, types.RetrievalPhase, bitswapCandidate))
+
+	// setup the linksystem to record bytes & blocks written -- since this isn't automatic w/o go-data-transfer
+	totalWritten := uint64(0)
+	blockCount := uint64(0)
+	cb := func(bytesWritten uint64) {
+		// record first byte received
+		if totalWritten == 0 {
+			br.events(events.FirstByte(br.request.RetrievalID, phaseStartTime, bitswapCandidate))
+		}
+		totalWritten += bytesWritten
+		blockCount++
```

Comment:
unsynchronized updates do feel a bit like a footgun waiting to happen, could we at least add a comment in here to remind us to synchronize these? Seems like a good spot to do some `atomic` operations.

Reply:
love a good atomic... cant hurt I guess.